packages/react-components/src/queries/useHistoricalAssetPropertyValues/requestExecution/getHistoricalAssetPropertyValueRequest.ts

```diff
@@ -1,33 +1,91 @@
 import {
   GetAssetPropertyValueHistory,
-  viewportEndDate,
-  viewportStartDate,
+  RequestResponse,
 } from '@iot-app-kit/core';
 import { HistoricalAssetPropertyValueRequest } from '../types';
+import { take } from 'lodash';
+import { mapViewport } from './mapViewport';
+import { mapTimeOrdering } from './mapTimeOrdering';
+
+const DEFAULT_MAX_RESULTS = 20000;
 
 export class GetHistoricalAssetPropertyValueRequest {
   readonly #getAssetPropertyValueHistory: GetAssetPropertyValueHistory;
-  readonly #maxResults = 20000;
+  private maxResults: number;
+  private valuesTargetNumber: number;
 
   constructor({
     getAssetPropertyValueHistory,
+    maxNumberOfValues = Infinity,
   }: {
     getAssetPropertyValueHistory: GetAssetPropertyValueHistory;
+    maxNumberOfValues?: number;
   }) {
     this.#getAssetPropertyValueHistory = getAssetPropertyValueHistory;
+    this.maxResults = Math.min(maxNumberOfValues, DEFAULT_MAX_RESULTS);
+    this.valuesTargetNumber = maxNumberOfValues;
   }
 
   public async send(
-    { viewport, ...request }: HistoricalAssetPropertyValueRequest,
+    { viewport, fetchMode, ...request }: HistoricalAssetPropertyValueRequest,
     settings: { abortSignal: AbortSignal }
   ) {
+    let nextToken = undefined;
+    const assetPropertyValueHistory: NonNullable<
+      RequestResponse<GetAssetPropertyValueHistory>['assetPropertyValueHistory']
+    > = [];
+
+    const timeOrdering = mapTimeOrdering({
+      fetchMode,
+      timeOrdering: request.timeOrdering,
+    });
+
+    const { startDate, endDate } = mapViewport({ fetchMode, viewport });
+
     try {
-      const startDate = viewport && viewportStartDate(viewport);
-      const endDate = viewport && viewportEndDate(viewport);
-      return await this.#getAssetPropertyValueHistory(
-        { startDate, endDate, maxResults: this.#maxResults, ...request },
-        settings
-      );
+      do {
+        settings.abortSignal.throwIfAborted();
+
+        const response: RequestResponse<GetAssetPropertyValueHistory> =
+          await this.#getAssetPropertyValueHistory(
+            {
+              ...request,
+              startDate,
+              endDate,
+              maxResults: this.maxResults,
+              timeOrdering,
+              nextToken,
+            },
+            settings
+          );
+
+        nextToken = response.nextToken;
+
+        const assetPropertyValues = response.assetPropertyValueHistory ?? [];
+        /**
+         * Only take the number of values determined by the target number
+         * The target number will be a running count from each paginated
+         * resposne.
+         */
+        const limitedAssetPropertyValues = take(
+          assetPropertyValues,
+          this.valuesTargetNumber
+        );
+        this.valuesTargetNumber -= assetPropertyValues.length;
+
+        /**
+         * Always return a list in ascending order
```

Comment:
Why always return list in ascending order?

Reply:
It will be important that the points are sorted in a consistent way so that we can do operations on them quickly. For instance, if we wanted to filter points only contained within the viewport, we can do it with binary search instead of running a standard filter operation.